File: backend/app/openrouter.py
# backend/app/openrouter.py
import json
from collections.abc import AsyncIterator
import httpx
from app.config import OPENROUTER_API_KEY, OPENROUTER_BASE_URL


async def stream_model_response(
    model_id: str,
    system_prompt: str,
    user_content: str,
    max_tokens: int = 2000,
) -> AsyncIterator[str]:
    """Stream a response from any model via OpenRouter.

    Yields text chunks as they arrive.
    """
    async with httpx.AsyncClient(timeout=120.0) as client:
        async with client.stream(
            "POST",
            OPENROUTER_BASE_URL,
            headers={
                "Authorization": f"Bearer {OPENROUTER_API_KEY}",
                "Content-Type": "application/json",
                "HTTP-Referer": "https://modelcouncil.ai",
                "X-Title": "Model Council",
            },
            json={
                "model": model_id,
                "messages": [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_content},
                ],
                "stream": True,
                "max_tokens": max_tokens,
            },
        ) as response:
            response.raise_for_status()
            async for line in response.aiter_lines():
                if not line.startswith("data: "):
                    continue
                data = line[6:]
                if data.strip() == "[DONE]":
                    break
                try:
                    chunk = json.loads(data)
                    delta = chunk.get("choices", [{}])[0].get("delta", {})
                    if text := delta.get("content"):
                        yield text
                except json.JSONDecodeError:
                    continue


async def call_model(
    model_id: str,
    system_prompt: str,
    user_content: str,
    max_tokens: int = 2000,
) -> str:
    """Call a model and return the complete response (non-streaming)."""
    chunks: list[str] = []
    async for chunk in stream_model_response(model_id, system_prompt, user_content, max_tokens):
        chunks.append(chunk)
    return "".join(chunks)

File: frontend/src/lib/openrouter.ts
const OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1/chat/completions";

export async function* streamModelResponse(
  modelId: string,
  systemPrompt: string,
  userContent: string,
  maxTokens: number = 2000,
): AsyncGenerator<string> {
  const apiKey = process.env.OPENROUTER_API_KEY;
  if (!apiKey) {
    throw new Error("OPENROUTER_API_KEY is not configured");
  }

  const response = await fetch(OPENROUTER_BASE_URL, {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${apiKey}`,
      "Content-Type": "application/json",
      "HTTP-Referer": "https://modelcouncil.ai",
      "X-Title": "Model Council",
    },
    body: JSON.stringify({
      model: modelId,
      messages: [
        { role: "system", content: systemPrompt },
        { role: "user", content: userContent },
      ],
      stream: true,
      max_tokens: maxTokens,
    }),
  });

  if (!response.ok) {
    const errorText = await response.text().catch(() => "");
    throw new Error(`OpenRouter ${response.status}: ${errorText || response.statusText}`);
  }

  const reader = response.body!.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (const line of lines) {
        if (!line.startsWith("data: ")) continue;
        const data = line.slice(6).trim();
        if (data === "[DONE]") return;
        try {
          const chunk = JSON.parse(data);
          const text = chunk.choices?.[0]?.delta?.content;
          if (text) yield text;
        } catch {
          continue;
        }
      }
    }
  } finally {
    reader.releaseLock();
  }
}
